### ISSUE TITLE
custom performance metric

### ISSUE THREAD

**Issue by ronencozen** [topepo/caret#552]
Hi,

I am trying to use a custom performance metric called "mapk",
while training an xgb model and getting the following error:

**Error in ctrl$summaryFunction(testOutput, lev, method) : 
  unused arguments (lev, method)**

Thanks,
Ronen

Here in the training code:

```
xgbModel <- train(
  clicked ~ . -display_id -ad_id,
  data = train,
  method = "xgbTree",
  metric = "mapk",
  maximize = TRUE,
  trControl = fitControl,
  tuneGrid = xgbGrid,
  verbose = 2)
```

and the function that implement the mapk metric:

```
fn_MAPK <- function (data, lev = c("No","Yes")) {
  data$obs <- factor(data$obs, levels=c(0,1), labels=c("No", "Yes"))
  data$pred <- factor(data$pred, levels=c(0,1), labels=c("No", "Yes"))
  eval_result <- data.frame(display_id=numeric(), score=numeric())
  for (i in unique(data$display_id)){
    df <- data %>% 
          filter(display_id == i)
    k <- nrow(df)
    actual <- df$obs
    predicted <- df$pred
    score <- mapk(k,actual,predicted)
    d <- c(i, score)
    eval_result <- rbind(eval_result, d) }
  out <- mean(eval_result[,2])
  names(out) <- "mapk"
  out}

apk <- function(k, actual, predicted)
 { score <- 0.0
   cnt <- 0.0
   for (i in 1:min(k,length(predicted)))
   { if (predicted[i] %in% actual && !(predicted[i] %in% predicted[0:(i-1)]))
     { cnt <- cnt + 1
       score <- score + cnt/i  }}
   score <- score / min(length(actual), k)
   score }  
  
 mapk <- function (k, actual, predicted)
 {if( length(actual)==0 || length(predicted)==0 ) 
   {return(0.0)}
    scores <- rep(0, length(actual))
   for (i in 1:length(scores))
   { scores[i] <- apk(k, actual[[i]], predicted[[i]]) }
   score <- mean(scores)
   score }
```
Here is an example that test the function, and shows that the function works as expected.
```
data<-data.frame(display_id=99,
              ad_id=c(1,2,3,4,5,6),
              obs=c(1,0,0,0,0,0),
              pred=c(0,0,0,0,0,1))
fn_MAPK(data)
```
 

>     mapk 
> 0.6666667

**Comment by topepo** [comment]
It looks like you need an argument `model` to `fn_MAPK`. See [this](https://topepo.github.io/caret/model-training-and-tuning.html#metrics) for the specifications

**Comment by ronencozen** [comment]
Hi ,
It seems that the missing model argument is not the issue, the problem was in the lines that convert the obs and pred variables into factor.
Now i face a different problem, the model runs and at the end issue the following error:

> Something is wrong; all the mapk metric values are missing:
>       mapk    
>  Min.   : NA  
>  1st Qu.: NA  
>  Median : NA  
>  Mean   :NaN  
>  3rd Qu.: NA  
>  Max.   : NA  
>  NA's   :1    
> Error in train.default(x, y, weights = w, ...) : Stopping
> In addition: Warning message:
> In nominalTrainWorkflow(x = x, y = y, wts = weights, info = trainInfo,  :
>   There were missing values in resampled performance measures.

The train set has no missing values, what do you think is wrong here.

Thanks,

Ronen

**Comment by ronencozen** [comment]
Hi,
Attached is the script and a small dataset to reproduce the error:
[train-err.txt](https://github.com/topepo/caret/files/644524/train-err.txt)
[small_train.txt](https://github.com/topepo/caret/files/644527/small_train.txt)

> R version 3.3.1 (2016-06-21)
> Platform: x86_64-w64-mingw32/x64 (64-bit)
> Running under: Windows >= 8 x64 (build 9200)
> 
> locale:
> [1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252
> [4] LC_NUMERIC=C                           LC_TIME=English_United States.1252    
> 
> attached base packages:
> [1] stats     graphics  grDevices utils     datasets  methods   base     
> 
> other attached packages:
> [1] xgboost_0.4-4   caret_6.0-71    ggplot2_2.1.0   lattice_0.20-33 dplyr_0.4.3     plyr_1.8.3      readr_0.2.2    
> 
> loaded via a namespace (and not attached):
>  [1] Rcpp_0.12.5        magrittr_1.5       splines_3.3.1      MASS_7.3-45        munsell_0.4.3      colorspace_1.2-6  
>  [7] R6_2.1.2           foreach_1.4.3      minqa_1.2.4        stringr_1.0.0      car_2.1-2          tools_3.3.1       
> [13] nnet_7.3-12        parallel_3.3.1     pbkrtest_0.4-6     grid_3.3.1         data.table_1.9.6   gtable_0.2.0      
> [19] nlme_3.1-128       mgcv_1.8-12        quantreg_5.24      DBI_0.4-1          MatrixModels_0.4-1 iterators_1.0.8   
> [25] lazyeval_0.1.10    lme4_1.1-12        assertthat_0.1     Matrix_1.2-6       nloptr_1.0.4       reshape2_1.4.1    
> [31] codetools_0.2-14   stringi_1.1.1      compiler_3.3.1     scales_0.4.0       stats4_3.3.1       SparseM_1.7       
> [37] chron_2.3-47

**Comment by topepo** [comment]
When you get the data for measuring performance, you only get the observed values, their predictions (and class probs, when requested), and the row ID. This is described [here](https://topepo.github.io/caret/model-training-and-tuning.html#metrics). For example: 
```
     pred obs        No        Yes rowIndex
1820  Yes  No 0.9060407 0.09395927      166
736    No  No 0.7769747 0.22302526     1640
753    No  No 0.2575246 0.74247543      283
2218   No  No 0.9849576 0.01504243     4001
880    No  No 0.5440227 0.45597735     3144
2623   No  No 0.2382688 0.76173121     4402
2159   No  No 0.7003100 0.29969000      831
3470   No  No 0.6347140 0.36528599     3202
1265  Yes  No 0.2350873 0.76491272     3140
3777  Yes Yes 0.5398311 0.46016892     3924
```
There is no `data$display_id` column so it doesn't work.  There is a feature request to have other columns come through to the performance calculation but that isn't trivial and will take a while. 

You _might_ be able to go this by joining the current results (in `data`)  using something like
```r
cbind(train[data$rowIndex,], data)
```
but you'll want to test this a lot. 

If you do, keep in mind that `train` does an initial evaluation of the summary function where a small subset of the data are scrambled (to get the variable names and do some tests). So the first execution of that function will be confusing some the observed classes in `data$obs` will not match your original data.

**Comment by ronencozen** [comment]
ok, thanks for the prompt reply!
